lib/Sema/CSDiag.cpp

```diff
@@ -4575,6 +4575,20 @@ bool FailureDiagnosis::visitSubscriptExpr(SubscriptExpr *SE) {
 
         if (CS.TC.getTypeOfExpressionWithoutApplying(expr, CS.DC, decl))
           return false;
+
+        // If we are down to a single candidate but with an unresolved
+        // index type, we can substitute in the base type to get a simpler
+        // and more concrete expected type for this subscript decl, in order
+        // to diagnose a better error.
+        if (baseType && indexType->hasUnresolvedType()) {
+          UncurriedCandidate cand = calleeInfo.candidates[0];
+          auto candType = baseType->getTypeOfMember(CS.DC->getParentModule(),
+                                                    cand.getDecl(), nullptr);
+          auto paramsType = candType->getAs<FunctionType>()->getInput();
```

Comment:
Looks good! Is it possible for candType to be something else but a function?

Reply:
No, at this point matching candidates definitely are subscripts, and have matching argument counts and matching types for any arguments that aren't unresolved.

Reply:
Great! I think this is good to go then!

Reply:
Thanks for working on this!

Reply:
Thanks for the reviews!